### PR TITLE
SAGE Test Cleanup First Pass

### PR DESCRIPTION
**Pull request by dtgreiner**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue. List any new dependencies)

This update should address the issues listed below. 

Cleanup for identified issues in SAGE TUP. 
- [X] Total in Q12 must equal total persons for Q5a
- [X] Language code in Q24D (21) is not a valid language response
- [X] 10D Row Totals
- [X] 16 Column Totals
- [X] 18 Column Totals
- [X] 22C Column Totals
- [X] 23D Column Totals
- [X] 23E Row Totals
- [X] 24A Column Totals
- [X] 25B Column Totals
- [X] 25J Column Totals
- [X] 27H Column Totals

Still outstanding
- [ ] You have an error in Q4a. There is an services-only project listed (project type 6), but "Affiliated with a residential project" is blank.

## Type of change
[//]: # (e.g., Bug fix, New feature, Code clean-up, Dependency update)
Bug fix

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

[//]: # NOTE: system tests may fail if there is no branch on the hmis-frontend that matches the Source  or Target branch of this PR. This is expected

### CODE REVIEW

File: app/models/concerns/hud_reports/incomes.rb [hunk 1]
```diff
@@ -203,14 +203,23 @@ module HudReports::Incomes
...
-                or(a_t["income_total_at_#{suffix}"].eq(nil))),
+              and(a_t["income_from_any_source_at_#{suffix}".to_sym].eq(nil).
+                or(a_t["income_from_any_source_at_#{suffix}"].eq(99)).
+                or(
+                  a_t["income_total_at_#{suffix}"].eq(nil).
+                  and(
+                    # Responses of 8 & 9 are expected to have total income nil.
+                    # These are filtered out to prevent duplicates. They are captured in a different row.
+                    a_t["income_from_any_source_at_#{suffix}"].not_eq(8).
+                    and(a_t["income_from_any_source_at_#{suffix}"].not_eq(9)),
```
Comment (eanders) [inline app/models/concerns/hud_reports/incomes.rb]: I think the following would have the same result and might be a bit easier to read
```suggestion
                    a_t["income_from_any_source_at_#{suffix}"].not_in([8, 9]),
```

File: app/models/concerns/hud_reports/start_to_move_in_question.rb [hunk 2]
```diff
@@ -68,11 +68,12 @@ def start_to_move_in_lengths
       [label, cond]
     end
 
+    # Make sure totals only include time to move in dates within the ranges being reported on
     ret.merge(
-      'Total (persons moved into housing)' => a_t[:hoh_move_in_date].between(@report.start_date..@report.end_date),
+      'Total (persons moved into housing)' => a_t[:hoh_move_in_date].between(@report.start_date..@report.end_date).and(a_t[:time_to_move_in].between(0..730)),
```
Comment (eanders) [inline app/models/concerns/hud_reports/start_to_move_in_question.rb]: is 730 a magic number (obviously 2 years) that maybe we should put in a constant?  If it's only used in these two places, it's probably fine as is.

File: drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_sixteen.rb [hunk 3]
```diff
@@ -67,7 +67,17 @@ def self.table_descriptions
...
-      not_collected = a_t["income_total_at_#{suffix}".to_sym].eq(nil)
+      not_collected = a_t["income_from_any_source_at_#{suffix}".to_sym].eq(nil).
+        or(a_t["income_from_any_source_at_#{suffix}".to_sym].eq(99)).
+        or(
+          a_t["income_total_at_#{suffix}"].eq(nil).
+          and(
+            # Responses of 8 & 9 are expected to have total income nil.
+            # These are filtered out to prevent duplicates. They are captured in a different row.
+            a_t["income_from_any_source_at_#{suffix}"].not_eq(8).
+            and(a_t["income_from_any_source_at_#{suffix}"].not_eq(9)),
```
Comment (eanders) [inline drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_sixteen.rb]: same comment as above if that works as expected.

File: drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_twenty_four.rb [hunk 4]
```diff
@@ -167,9 +167,9 @@ def q24d_language_of_persons_requiring_translation_assistance
         'Able to maintain the housing they had at project start--Only with financial assistance other than a subsidy' => a_t[:housing_assessment].eq(1).
           and(a_t[:subsidy_information].eq(4)),
         'Moved to new housing unit--With on-going subsidy' => a_t[:housing_assessment].eq(2).
-          and(a_t[:subsidy_information].eq(3)),
+          and(a_t[:subsidy_information].eq(11)),
```
Comment (eanders) [inline drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_twenty_four.rb:170]: These are fun!

**Review by eanders — APPROVED** [review]
Couple possible clarifications, but otherwise this looks great, thanks @dtgreiner !

File: app/models/concerns/hud_reports/incomes.rb (hunk 1 above)
Comment (dtgreiner) [reply app/models/concerns/hud_reports/incomes.rb]: This has been updated in [this commit](https://github.com/greenriver/hmis-warehouse/pull/5000/commits/6d299d090919f6b3640dbb18b822a4c0eae6593c)

File: drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_sixteen.rb (hunk 3 above)
Comment (dtgreiner) [reply drivers/hud_apr/app/models/hud_apr/generators/shared/fy2024/question_sixteen.rb]: This has been updated in [this commit](https://github.com/greenriver/hmis-warehouse/pull/5000/commits/6d299d090919f6b3640dbb18b822a4c0eae6593c)

File: app/models/concerns/hud_reports/start_to_move_in_question.rb (hunk 2 above)
Comment (dtgreiner) [reply app/models/concerns/hud_reports/start_to_move_in_question.rb]: I pulled the number into to a variable with a better description. This is the only place it is being used. I could probably pull it from the arel query being returned from the lengths method being called a bit above, but that was getting a bit messy and less readable so I just set it to 730 with a descriptive name and comment.

**Comment by dtgreiner** [comment]
@eanders This is ready for another look when you have the chance. Thanks!

File: app/models/concerns/hud_reports/start_to_move_in_question.rb (hunk 2 above)
Comment (eanders) [reply app/models/concerns/hud_reports/start_to_move_in_question.rb]: Much more readable.  Thank you!

**Pull request merged**